core/ledger/kvledger/txmgmt/statedb/stateleveldb/stateleveldb.go

```diff
@@ -268,3 +284,64 @@ func (scanner *kvScanner) GetBookmarkAndClose() string {
 	scanner.Close()
 	return retval
 }
+
+type fullDBScanner struct {
+	db               *leveldbhelper.DBHandle
+	dbItr            iterator.Iterator
+	skipNamespace    func(string) bool
+	currentNamespace string
+}
+
+func newFullDBScanner(db *leveldbhelper.DBHandle, skipNamespace func(string) bool) (*fullDBScanner, byte, error) {
+	dbItr := db.GetIterator(dataKeyPrefix, dataKeyStopper)
+	if err := dbItr.Error(); err != nil {
+		return nil, byte(0), errors.Wrap(err, "internal leveldb error while obtaining db iterator")
+	}
+	return &fullDBScanner{
+			db:            db,
+			dbItr:         dbItr,
+			skipNamespace: skipNamespace,
+		},
+		fullScanIteratorValueFormat,
+		nil
+}
+
+// Next returns the key-values in the lexical order of <Namespace, key>
+// The bytes returned for the <version, value, metadata> are the same as they are stored in the leveldb.
+// Since, the primary intended use of this function is to generate the snapshot files for the statedb, the same
+// bytes can be consumed back as is. Hence, we do not decode or transform these bytes for the efficiency
+func (s *fullDBScanner) Next() (*statedb.CompositeKey, []byte, error) {
+	for s.dbItr.Next() {
+		dbKey := s.dbItr.Key()
+		logger.Debugw("", "dbKey", dbKey)
+		dbVal := s.dbItr.Value()
+		ns, key := decodeDataKey(dbKey)
+		compositeKey := &statedb.CompositeKey{
+			Namespace: ns,
+			Key:       key,
+		}
+
+		switch {
+		case ns == s.currentNamespace:
+			return compositeKey, dbVal, nil
+
+		// new namespace begins
+		case !s.skipNamespace(ns):
+			s.currentNamespace = ns
+			return compositeKey, dbVal, nil
+
+		// skip the new namespace
+		default:
+			s.dbItr.Release()
+			s.dbItr = s.db.GetIterator(dataKeyStarterForNextNamespace(ns), dataKeyStopper)
```

Comment:
can't we use `dbItr.Seek()` instead?

Reply:
Yes, that would have been an obvious choice but that requires some work in the `leveldbhelper` package. That package returns the iterator returned by the leveldb as is. But for `Seek` to function, we would need to append the channel name. We did not need it so far, as we always scanned with in a namespace.

Having said that, it would be good to write a wrapper iterator in the `leveldbhelper` package and override the `Seek` function and return that instead. I'll do that in a separate PR and will use that here.

Reply:
I see. I wasn't aware of this. I just looked at the [GetIterator()](https://github.com/hyperledger/fabric/blob/master/common/ledger/util/leveldbhelper/leveldb_provider.go#L175) and noticed that it appends the channel name. Now I understand what you meant above. Good to know about this wrapper. Thanks.